src/quicktype-core/language/CPlusPlus.ts

```diff
@@ -971,99 +1036,209 @@ export class CPlusPlusRenderer extends ConvenienceRenderer {
+    protected updatePropertyTypes(recurseIntoUnion:boolean, level:number, includes: IncludeMap, propertyTypes: Map<string, IncludeRecord>, proposedIncludeKind:IncludeKind|undefined, typeName:string, t:Type, defName:string): void {
+        let propRecord: IncludeRecord = { kind: undefined, objectType: undefined };
+
+        if (t instanceof ClassType) {
+            /** 
+             * Ok. We can NOT forward declare direct class members, e.g. a class type is included
+             * at level#0. HOWEVER if it is not a direct class member (e.g. std::shared_ptr<Class>),
+             * - level > 0 - then we can SURELY forward declare it.
+             */
+            propRecord.objectType = ObjectType.Class;
+            propRecord.kind = proposedIncludeKind !== undefined ? proposedIncludeKind : level === 0 ? IncludeKind.Include : IncludeKind.ForwardDeclare;
+        } else if (t instanceof EnumType) {
+            propRecord.objectType = ObjectType.Enum;
+            propRecord.kind = proposedIncludeKind !== undefined ? proposedIncludeKind : IncludeKind.ForwardDeclare;
+        } else if (t instanceof UnionType) {
+            /** Recurse into the union */
+            const [maybeNull, nonNulls] = removeNullFromUnion(t, true);
+            if (recurseIntoUnion) {
+                if (nonNulls !== undefined) {
+                    for (const tt of nonNulls) {
+                        /**
+                         * We can forward declare ANYTHING in a union, however
+                         * variants we need to include directly (they are typedefs)
+                         */
+                        this.updateIncludes(maybeNull === undefined, level + 1, includes, tt, defName);
+                    }
                 }
 
+                return;
+            }
+
+            /** 
+             * Even if forward declaration is possible, 
+             * but we don't want to check every possible union
+             * member types, simple include the definition.
+             */
+            propRecord.objectType = ObjectType.Union;
+            propRecord.kind = IncludeKind.Include;
+        }
+
+        if (!propertyTypes.has(typeName)) {
+            propertyTypes.set( typeName, propRecord);
+        } else {
+            /** 
+             * We don't care what we have already as a includekind
+             * for the type, we only update this if we want to
+             * set a stronger kind
+             */
+            if (propRecord.kind === IncludeKind.Include) {
+                propertyTypes.set( typeName, propRecord);
+            }
+        }
+    }
+
+    /** 
+     * if forwardDeclare is set to true we force it, level shows how 
+     * deep we are in in the definition
+     */
+    protected updateIncludes(forwardDeclare: boolean | undefined, level:number, includes: IncludeMap, propertyType: Type, defName:string): void {
+        /** God we don't really have custom sets */
```

Comment:
Custom sets?

Reply:
Meh, coming from a C++ world, where I can overload an operator so that I can create a class which will be properly sorted in the Set by my own definition I am quite missing that here, or has the JavaScript/TypeScript land evolved that much that equals() can be overridden somehow?

Reply:
Oh, no, I find it very unfortunate that `Map` and `Set` don't have configurable hashing and equality.  In particular, you can't do value equality, only what `===` does.  We do need it in quicktype in at least one place, which is why I created [this monstrosity](https://github.com/quicktype/collection-utils/blob/master/index.ts#L427)